tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/FunCtorRule.scala

```diff
@@ -101,28 +99,20 @@ class FunCtorRule(rewriter: SymbStateRewriter) extends RewritingRule {
       state: SymbState,
       mapEx: TlaEx,
       varName: String,
-      setEx: TlaEx,
-      oldCells: Seq[ArenaCell]) = {
-    // similar to SymbStateRewriter.rewriteSeqUntilDone and SetFilterRule
-    def process(st: SymbState, seq: Seq[ArenaCell]): (SymbState, Seq[TlaEx]) = {
-      seq match {
-        case Seq() =>
-          (st, List())
+      oldCells: Seq[ArenaCell]): (SymbState, Seq[ArenaCell]) = {
+    var nextState = state
 
-        case cell :: tail =>
-          val (ts: SymbState, nt: List[TlaEx]) = process(st, tail)
-          val newBinding = Binding(ts.binding.toMap + (varName -> cell))
-          val cellState = new SymbState(mapEx, ts.arena, newBinding)
-          // add [cell/x]
-          val ns = rewriter.rewriteUntilDone(cellState)
-          (ns.setBinding(ts.binding), ns.ex :: nt) // reset binding and add the new expression in the head
-      }
+    // similar to SymbStateRewriter.rewriteSeqUntilDone and SetFilterRule
+    def mapOne(cell: ArenaCell): ArenaCell = {
+      // rewrite mapEx[cell/varName]
+      val newBinding = Binding(nextState.binding.toMap + (varName -> cell))
+      nextState = rewriter.rewriteUntilDone(nextState.setRex(mapEx).setBinding(newBinding))
+      nextState.asCell
     }
 
     // compute mappings for all the cells (some of them may coincide, that's fine)
-    val (newState: SymbState, newEs: Seq[TlaEx]) = process(state, oldCells)
-    val newCells = newEs.map(newState.arena.findCellByNameEx)
+    val mappedCells = oldCells.map(mapOne)
     // keep the sequence of results as it is, as it is will be zipped by the function constructor
-    (newState, newCells)
+    (nextState.setBinding(state.binding), mappedCells)
```

Comment:
Can we make `nextState` immutable? I don't expect comprehensions to have side-effects.

Reply:
We can, but this will make the code much harder to read. We have this mix of mutable modifications to the state and the arena in many places, and it is significantly shorter and easier to read than the pure folds.

Reply:
But in this case you could just return the state of each step; then `nextState` is the tail of that `Seq` and you can `.map(_.asCell)` to get `newCells`?

Reply:
ah, yes, this could work, unless the sequence is empty

Reply:
But you're right, I missed the empty Seq corner case. Resolve this as you see fit!